app/providers/atmosphere/optimization_strategy/default.rb

```diff
@@ -64,41 +86,67 @@ def preferences
         end
       end
 
-      class VmAndFlavor
+      class VmAndFlavorAndTenant
         include Atmosphere::Utils
 
-        def initialize(tmpls, options={})
+        def initialize(tmpls, appliance, options={})
           @tmpls = tmpls
+          @appliance = appliance
           @options = options
         end
 
         def select
-          # TODO: Hack. Assumes all tenants in a given CS share the same set of flavors.
-          opt_flavors_and_tmpls_map = tmpls.inject({}) do |hsh, tmpl|
-            opt_fl = (
-              min_elements_by(
-                tmpl.tenants.first.virtual_machine_flavors.active.select do |f|
-                  f.supports_architecture?(tmpl.architecture) &&
-                  f.memory >= min_mem &&
-                  f.cpu >= min_cpu &&
-                  f.hdd >= min_hdd
-                end
-              ) {|f| tmpl.get_hourly_cost_for(f) || Float::INFINITY }
-            ).sort!{ |x,y| y.memory <=> x.memory }.last
-            hsh[opt_fl] = tmpl if opt_fl
-            hsh
+          # We are given a list of tmpls which MAY be spawned by the user in the context of @appliance.
+          # We are asked to return a single template, a specific flavor and a specific tenant for instantiation.
+          # Flavor prices may vary by tenant.
+          # !CAUTION! @appliance can be nil -- this will happen when the method is invoked prior to
+          # creation of an appliance.
+          best_template = nil
+          best_tenant = nil
+          best_flavor = nil
+          instantiation_cost = Float::INFINITY
+
+          tmpls.each do |tmpl|
+            # Assuming tmpl is selected, figure out which tenant and which flavor to use.
+            candidate_tenants = get_candidate_tenants_for_template(tmpl)
+            # For each candidate tenant find optimal flavor and the corresponding VM instantiation cost
+            # Note that candidate_tenants can be blank - we must handle this case.
+            opt_tenant = nil
+            opt_flavor_and_cost = [nil, Float::INFINITY]
```

Comment:
Maybe hash or struct with meaningful keys would be more understandable than `opt_flavor_and_cost[1]`?

E.g.

```ruby
opt_flavor_and_cost = { flavor: nil, price: Float::INFINITY }
# or
opt_flavor_and_cost = Struct.new(:flavor, :price).new(nil, Float::INFINITY)
```

and then:

```ruby
opt_flavor_and_cost[:flavor]
opt_flavor_and_cost[:price]
#or
opt_flavor_and_cost.flavor
opt_flavor_and_cost.price
```